README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# svgpatternsimple <img src="man/figures/logo.svg" align="right" height=230/>

<!-- badges: start -->

![](http://img.shields.io/badge/cool-useless-green.svg)
![](http://img.shields.io/badge/mini-verse-blue.svg)
<!-- badges: end -->

`svgpatternsimple` provides a basic set of simple repeating SVG
patterns.

For online documentation, see the [svgpatternsimple pkgdown
website](https://coolbutuseless.github.io/package/svgpatternsimple/index.html).

## Installation

You can install from
[GitHub](https://github.com/coolbutuseless/svgpatternsimple) with:

``` r
# install.packages("devtools")
install_github("coolbutuseless/poissoned")        # Generate points via poisson disk sampling
install_github("coolbutuseless/minnisvg")         # Build SVG documents with R
install_github("coolbutuseless/svgpatternsimple") # This package
```

## Example: What is a pattern?

The patterns generated by this package are SVG `<pattern>` elements
created with [`minisvg`](https://github.com/coolbutuseless/minisvg).

`minisvg` elements are R6 objects which render to a string representing
SVG.

To wrap the pattern in a generic SVG document use the method
`$as_full_svg()`, and to display it in an RStudio session use `$show()`

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create a dot pattern
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dot <- create_pattern_dot(id = 'dot')
```

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# All `minisvg::SVGPattern` objects have a built in '$show()' method.
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dot$show()
```

<pre><details><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;svg viewBox="0 0 400 400" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"&gt;
  &lt;defs&gt;
    &lt;pattern id="dot" width="20" height="20" patternTransform="rotate(45 0 0)" patternUnits="userSpaceOnUse"&gt;
      &lt;rect width="100%" height="100%" style="fill:#ffffff; fill-opacity:1; stroke:none;" /&gt;
      &lt;circle cx="10" cy="10" r="1.9" style="stroke:#000000; fill:#000000; fill-opacity:1; stroke-opacity:1;" /&gt;
    &lt;/pattern&gt;
  &lt;/defs&gt;
  &lt;rect x="0" y="0" width="100%" height="100%" style="fill: url(#dot) #fff;" /&gt;
&lt;/svg&gt;</details></pre>

<img src="man/figures/README-example-dot.svg" />

## Applying a pattern within an SVG document

1.  Create a `minisvg` document
2.  Cretae a pattern
3.  Add the pattern definition to the SVG document
4.  Create a rectangle
5.  Assign the pattern as the fill for the rectangle
6.  Append the rectangle to the document.

<!-- end list -->

``` r
library(minisvg)
doc <- minisvg::svg_doc(width=400, height=100)

mypattern <- create_pattern_hex(id = 'hexpattern')
doc$defs(mypattern)

myrect <- stag$rect(x=0, y=0, width=400, height=100, fill=mypattern)

doc$append(myrect)

# doc$show()
```

<svg width="400" height="100" viewBox="0 0 400 100" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">

<defs>
<pattern id="hexpattern" width="34.6410161513775" height="60" patternTransform="rotate(0 0 0)" patternUnits="userSpaceOnUse">
<rect width="100%" height="100%" style="fill:#ffffff; fill-opacity:1; stroke:none;" />
<g style="stroke:#000000; stroke-width:1; stroke-opacity:1;">
<line x1="17.3205080756888" y1="10" x2="0" y2="20" />
<line x1="17.3205080756888" y1="10" x2="34.6410161513775" y2="20" />
<line x1="17.3205080756888" y1="50" x2="0" y2="40" />
<line x1="17.3205080756888" y1="50" x2="34.6410161513775" y2="40" />
<line x1="17.3205080756888" y1="0" x2="17.3205080756888" y2="10" />
<line x1="17.3205080756888" y1="50" x2="17.3205080756888" y2="60" />
<line x1="0" y1="20" x2="0" y2="40" />
<line x1="34.6410161513775" y1="20" x2="34.6410161513775" y2="40" />
</g> </pattern> </defs>
<rect x="0" y="0" width="400" height="100" fill="url('#hexpattern')" />

</svg>

## Example: All patterns

There are currently 7 pattern types included in this package.  
Patterns are created using `create_[type]_pattern()`, and arguments
control the appearance of the generated pattern e.g. opacity,
orientation etc.

``` r
dot          <- create_pattern_dot     (id = 'p1' , fill_fraction = 0.4)
stripe       <- create_pattern_stripe  (id = 'p2' , angle = 60)
hatch        <- create_pattern_hatch   (id = 'p3' , spacing = 50)
stipple      <- create_pattern_stipple (id = 'p4' , colour = 'hotpink')
hex          <- create_pattern_hex     (id = 'p5' , fill_fraction = 0.1)
check        <- create_pattern_check   (id = 'p6' , spacing = 30, colour = 'darkgreen')
gradient     <- create_pattern_gradient(id = 'p7' , colour1 = 'hotpink', colour2 = 'darkblue')

pattern_list <- list(dot, stripe, hatch, stipple, hex, check, gradient)
doc <- SVGPatternList_to_svg(pattern_list, height = 100, ncol = 2)
# doc$show()
doc$save("man/figures/README-all-patterns.svg")
```

<pre><details><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;svg width="800" height="400" viewBox="0 0 800 400" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"&gt;
  &lt;defs&gt;
    &lt;pattern id="p1" width="20" height="20" patternTransform="rotate(45 0 0)" patternUnits="userSpaceOnUse"&gt;
      &lt;rect width="100%" height="100%" style="fill:#ffffff; fill-opacity:1; stroke:none;" /&gt;
      &lt;circle cx="10" cy="10" r="3.81" style="stroke:#000000; fill:#000000; fill-opacity:1; stroke-opacity:1;" /&gt;
    &lt;/pattern&gt;
    &lt;pattern id="p2" width="20" height="20" patternTransform="rotate(60 0 0)" patternUnits="userSpaceOnUse"&gt;
      &lt;rect width="100%" height="100%" style="fill:#ffffff; fill-opacity:1; stroke:none;" /&gt;
      &lt;line x1="0" y1="0" x2="0" y2="20" style="stroke:#000000; stroke-width:8; stroke-opacity:1;" /&gt;
    &lt;/pattern&gt;
    &lt;pattern id="p3" width="50" height="50" patternTransform="rotate(45 0 0)" patternUnits="userSpaceOnUse"&gt;
      &lt;rect width="100%" height="100%" style="fill:#ffffff; fill-opacity:1; stroke:none;" /&gt;
      &lt;g style="stroke:#000000; stroke-width:20; stroke-opacity:1;"&gt;
        &lt;line x1="0" y1="0" x2="0" y2="50" /&gt;
        &lt;line x1="0" y1="0" x2="50" y2="0" /&gt;
      &lt;/g&gt;
    &lt;/pattern&gt;
    &lt;pattern id="p4" width="400" height="400" patternUnits="userSpaceOnUse"&gt;
      &lt;rect width="100%" height="100%" style="fill:#ffffff; fill-opacity:1; stroke:none;" /&gt;
      &lt;g&gt;
        &lt;circle cx='-7.22' cy='-4.32' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='-15.61' cy='47.93' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='-16.54' cy='78.06' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='-7.22' cy='395.68' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='18.12' cy='10.17' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='13.26' cy='61.42' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='9.46' cy='96.27' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='5.75' cy='131.56' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='4.56' cy='197.36' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='3.19' cy='235.5' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='4.14' cy='282.68' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='17.42' cy='328.23' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='18.17' cy='378.64' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='18.12' cy='410.17' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='22.23' cy='38.31' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='38.96' cy='102.91' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='24.14' cy='174.25' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='28.66' cy='220.65' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='28.52' cy='260.29' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='29.89' cy='301.87' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='32.04' cy='353.3' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='44.01' cy='16.75' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='49.56' cy='61.2' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='51.86' cy='129.59' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='54.86' cy='206.89' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='54.21' cy='379.55' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='44.01' cy='416.75' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='66.79' cy='34.14' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='69.35' cy='154.15' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='67.29' cy='239.33' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='61.73' cy='278.56' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='67.13' cy='306.66' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='79.8' cy='342.52' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='92.56' cy='17.08' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='95.83' cy='68.24' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='84.71' cy='103.95' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='97.92' cy='152.81' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='91.75' cy='191.84' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='97.77' cy='228.05' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='90.21' cy='258.89' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='86.8' cy='379.54' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='92.56' cy='417.08' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='115.3' cy='-17.66' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='113.89' cy='41.27' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='119.3' cy='124.51' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='109.95' cy='279.92' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='103.78' cy='325.47' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='115.3' cy='382.34' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='129.24' cy='16.23' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='129.91' cy='65.66' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='126' cy='97.26' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='124.3' cy='176.53' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='132.86' cy='203.55' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='120.95' cy='231.21' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='139.66' cy='302.33' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='120.57' cy='352.94' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='129.24' cy='416.23' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='159.36' cy='40.5' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='144.52' cy='128.98' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='153.1' cy='174.42' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='143.23' cy='263.88' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='156.87' cy='335.79' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='169.19' cy='19.82' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='173.81' cy='72.67' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='166.07' cy='105.27' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='179.33' cy='135.92' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='160.9' cy='207.54' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='170.61' cy='235.55' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='176.37' cy='312.63' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='161.62' cy='378.23' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='169.19' cy='419.82' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='199.15' cy='20' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='189.35' cy='44.19' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='181.88' cy='168.91' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='191.13' cy='212.72' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='188.58' cy='263.27' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='188.12' cy='339.77' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='195.85' cy='364.11' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='199.15' cy='420' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='213.82' cy='-5.9' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;
&lt;circle cx='203.03' cy='74.37' r='4' style='fill: hotpink; stroke: hotpink; fill-opacity: 1; stroke-opacity: 1' /&gt;

NEWS.md

# svgpatternsimple 0.2.0  2020-02-01

* Remove old `{devoutsvg-0.1.0}` compatibility.

# svgpatternsimple 0.1.1

* Improved generation of pattern palettes
* Stipple style now working

# svgpatternsimple 0.1.0

* Initial release


R/api-core.R



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# How bits are packed for simple patterns
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
all_pack_specs <- list(
  name_only = list(
    pattern_name   = list(type = 'choice' , nbits =  3, options = valid_pattern_names),
    padding        = list(type = 'integer', nbits = 29)
  ),
  stripe = list(
    pattern_name   = list(type = 'choice' , nbits =  3, options = valid_pattern_names),
    angle          = list(type = 'custom' , nbits =  2, pack_func = ~round((.x %% (3.5/4*180)) / 45), unpack_func = ~.x * 45),
    spacing        = list(type = 'scaled' , nbits =  4, max = 255),
    fill_fraction  = list(type = 'scaled' , nbits =  4, max = 1),
    fg_alpha       = list(type = 'scaled' , nbits =  3, max = 1),
    colour         = list(type = 'colour' , nbits =  8),
    alpha          = list(type = 'scaled' , nbits =  8, max = 1)
  ),
  dot = list(
    pattern_name   = list(type = 'choice' , nbits =  3, options = valid_pattern_names),
    angle          = list(type = 'custom' , nbits =  2, pack_func = ~round((.x %% (3.5/4*90)) / 22.5), unpack_func = ~.x * 22.5),
    spacing        = list(type = 'scaled' , nbits =  4, max = 255),
    fill_fraction  = list(type = 'scaled' , nbits =  4, max = 1),
    fg_alpha       = list(type = 'scaled' , nbits =  3, max = 1),
    colour         = list(type = 'colour' , nbits =  8),
    alpha          = list(type = 'scaled' , nbits =  8, max = 1)
  ),
  stipple =  list(
    pattern_name   = list(type = 'choice' , nbits =  3, options = valid_pattern_names),
    colour         = list(type = 'colour' , nbits =  8),
    spacing        = list(type = 'scaled' , nbits =  5, max = 255),
    fill_fraction  = list(type = 'scaled' , nbits =  4, max = 1),
    fg_alpha       = list(type = 'scaled' , nbits =  4, max = 1),
    alpha          = list(type = 'scaled' , nbits =  8, max = 1)
  ),
  gradient = list(
    pattern_name   = list(type = 'choice' , nbits =  3, options = valid_pattern_names),
    angle          = list(type = 'scaled' , nbits =  5, max = 360, cyclical = TRUE),
    colour1        = list(type = 'colour' , nbits =  8),
    colour2        = list(type = 'colour' , nbits =  8),
    alpha          = list(type = 'scaled' , nbits =  8, max = 1)
  ),
  check = list(
    pattern_name   = list(type = 'choice' , nbits =  3, options = valid_pattern_names),
    angle          = list(type = 'scaled' , nbits =  6, max = 90, cyclical = TRUE),
    spacing        = list(type = 'scaled' , nbits =  4, max = 255),
    fg_alpha       = list(type = 'scaled' , nbits =  3, max = 1),
    colour         = list(type = 'colour' , nbits =  8),
    alpha          = list(type = 'scaled' , nbits =  8, max = 1)
  ),
  null = list(
    pattern_name   = list(type = 'choice' , nbits =  3, options = valid_pattern_names),
    colour         = list(type = 'colour' , nbits =  21),
    alpha          = list(type = 'scaled' , nbits =  8, max = 1)
  )
)

all_pack_specs[['hatch'  ]] <- all_pack_specs[['dot']]
all_pack_specs[['hex'    ]] <- all_pack_specs[['dot']]


all_defaults <- list(
  stripe = list(
    pattern_name  = 'stripe',
    angle         = 45,
    spacing       = 20,
    fill_fraction = 0.2,
    fg_alpha      = 1,
    colour        = '#000000',
    alpha         = 1
  ),
  dot = list(
    pattern_name  = 'dot',
    angle         = 22.5,
    spacing       = 20,
    fill_fraction = 0.2,
    fg_alpha      = 1,
    colour        = '#000000',
    alpha         = 1
  ),
  hatch = list(
    pattern_name  = 'hatch',
    angle         = 22.5,
    spacing       = 20,
    fill_fraction = 0.2,
    fg_alpha      = 1,
    colour        = '#000000',
    alpha         = 1
  ),
  hex = list(
    pattern_name  = 'hex',
    angle         = 22.5,
    spacing       = 20,
    fill_fraction = 0.2,
    fg_alpha      = 1,
    colour        = '#000000',
    alpha         = 1
  ),
  stipple = list(
    pattern_name  = 'stipple',
    spacing       = 20,
    fill_fraction = 0.2,
    fg_alpha      = 1,
    colour        = '#000000',
    alpha         = 1
  ),
  gradient = list(
    pattern_name = 'gradient',
    angle        = 45,
    colour1      = '#00204D',
    colour2      = '#FFEA46',
    alpha        = 1
  ),
  check = list(
    pattern_name  = 'check',
    id            = NULL,
    angle         = 45,
    spacing       = 20,
    fg_alpha      = 1.0,
    colour        = '#000000',
    alpha         = 1.0
  ),
  null = list(
    colour = '#ffffff',
    alpha  = 1
  )
)





#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Is the rgba_vec interpretable as a pattern?
#'
#' @param rgba_vec 4 element vector of RGBA values, each in the range [0, 255]
#'
#' @return logical value indicating whether or not thig RGBA colour is
#'         interpretable as a pattern in this package?
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
is_valid_pattern_encoding <- function(rgba_vec) {
  is_black <- function(vec) {
    vec[1] == 0 && vec[2] == 0 && vec[3] == 0
  }

  is_white <- function(vec) {
    vec[1] == 255 && vec[2] == 255 && vec[3] == 255
  }

  is_transparent <- function(vec) {
    vec[4] == 0
  }


  !(is_black(rgba_vec) || is_white(rgba_vec) || is_transparent(rgba_vec))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Encode a pattern as a hex colour
#'
#' @param ... pattern parameters
#'
#' @return Return a hex colour representing the parameters
#'
#' @import lofi
#' @importFrom utils modifyList
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
encode_pattern_params_as_hex_colour <- function(...) {
  params <- list(...)
  if (!'pattern_name' %in% names(params)) {
    stop("'svgpatternsimple' needs a 'pattern_name' argument. None given.", call. = FALSE)
  }

  pattern_name <- params$pattern_name
  pack_spec    <- all_pack_specs[[pattern_name]]
  defaults     <- all_defaults[[pattern_name]]

  values <- modifyList(defaults, params)

  int32  <- lofi::pack(values, pack_spec)
  lofi::int32_to_hex_colour(int32)
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create the pattern id for a given hex colour for this package
#'
#' @param rgba_vec 4 element RGBA colour vector. Each element in range [0, 255]
#'
#' @return character string to use as a pattern ID
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
create_pattern_id_from_rgba_vec <- function(rgba_vec) {
  hex_colour <- lofi::rgba_vec_to_hex_colour(rgba_vec)
  paste0("simple-", substr(hex_colour, 2, 9))
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Decode a 4-element RGBA colour vector as an SVG pattern
#'
#' @param rgba_vec 4 element RGBA colour vector. Each element in range [0, 255]
#'
#' @return svg_pattern object
#'
#' @import lofi
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
decode_pattern_from_rgba_vec <- function(rgba_vec) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Need an int32 so we can unpack
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  hex_colour <- lofi::rgba_vec_to_hex_colour(rgba_vec)
  int32      <- lofi::rgba_vec_to_int32(rgba_vec)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Figure out which pattern this is by unpacking the name only
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  name_only_pack_spec <- all_pack_specs[['name_only']]
  pattern_name        <- lofi::unpack(int32, name_only_pack_spec)[['pattern_name']]

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Select the correct pattern pack_spec and unpack the int32
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  pack_spec <- all_pack_specs[[pattern_name]]
  params    <- lofi::unpack(int32, pack_spec)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Ensure we have a sane ID attached to this pattern
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  params$id  <- create_pattern_id_from_rgba_vec(rgba_vec)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Select the function for creating the pattern
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  f <- switch(
    params$pattern_name,
    stripe   = create_stripe_pattern  ,
    hatch    = create_hatch_pattern   ,
    dot      = create_dot_pattern     ,
    stipple  = create_stipple_pattern ,
    hex      = create_hex_pattern     ,
    null     = create_null_pattern    ,
    gradient = create_gradient_pattern,
    check    = create_check_pattern   ,
    stop("decode_pattern_from_hex_colour(): Unknown pattern_name: ", deparse(params$pattern_name), call. = FALSE)
  )


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Create the pattern from the params and return it
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  do.call(f, params)
}





